src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5532,6 +5532,107 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
         cbnzw(tmp5, DONE);
       }
     }
+  } else if (UseCompactObjectHeaders) {
+    Label NEXT_64BYTE, NEXT_16BYTE, TAIL, TAIL03, TAIL01;
+    mov(result, false);
+    // if (a1 == null || a2 == null)
+    //     return false;
```

Comment:
Is there a reference implementation for this null check?

One of the existing implementation uses:
    tst(a1, a2);
which I think is quite clever; the other one uses:
    cbz(a1, DONE);
    ldrw(cnt1, Address(a1, length_offset));
    cbz(a2, DONE);
    ldrw(cnt2, Address(a2, length_offset));
    cmpw(cnt2, cnt1);
which moves ldrw(cnt1) ahead. My guess is that cmpw depends on the availability of cnt1 so it wants to load cnt1 sooner. Is there any evaluation that suggests the proposed implementation is superior?
    cbz(a1, DONE);
    cbz(a2, DONE);
    ldrw(cnt1, Address(a1, length_offset));
    ldrw(cnt2, Address(a2, length_offset));
    cmpw(cnt2, cnt1);

Reply:
done